



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="Project documentation for AEGIS">
      
      
        <link rel="canonical" href="http://aegis.stanford.edu/api/">
      
      
        <meta name="author" content="Junjie (Jason) Zhu">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0, mkdocs-material-3.0.4">
    
    
      
        <title>API Documentation - AEGIS Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.451f80e5.css">
      
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../css/extra.css">
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="../#python-documentation" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="http://aegis.stanford.edu" title="AEGIS Documentation" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                AEGIS Documentation
              </span>
              <span class="md-header-nav__topic">
                API Documentation
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/junjiezhujason/aegis" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="http://aegis.stanford.edu" title="AEGIS Documentation" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    AEGIS Documentation
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/junjiezhujason/aegis" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" checked>
    
    <label class="md-nav__link" for="nav-2">
      User Guide
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        User Guide
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../install/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../tutorial/" title="Video Tutorials" class="md-nav__link">
      Video Tutorials
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../manual/" title="Web Interface Manual" class="md-nav__link">
      Web Interface Manual
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        API Documentation
      </label>
    
    <a href="./" title="API Documentation" class="md-nav__link md-nav__link--active">
      API Documentation
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#serverdagraphpy" title="server/dagraph.py" class="md-nav__link">
    server/dagraph.py
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#serveralgorithmspy" title="server/algorithms.py" class="md-nav__link">
    server/algorithms.py
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3">
    
    <label class="md-nav__link" for="nav-3">
      Jupyter Notebooks
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        Jupyter Notebooks
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../notebooks/aegis_single_cell_example.html" title="Single Cell RNA-seq Application" class="md-nav__link">
      Single Cell RNA-seq Application
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../notebooks/aegis_benchmark_part1.html" title="Benchmark for Testing: Part 1" class="md-nav__link">
      Benchmark for Testing: Part 1
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../notebooks/aegis_benchmark_part2.html" title="Benchmark for Testing: Part 2" class="md-nav__link">
      Benchmark for Testing: Part 2
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#serverdagraphpy" title="server/dagraph.py" class="md-nav__link">
    server/dagraph.py
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#serveralgorithmspy" title="server/algorithms.py" class="md-nav__link">
    server/algorithms.py
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="python-documentation">Python Documentation</h1>
<h3 id="serverdagraphpy">server/dagraph.py</h3>
<div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">BaseNode</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Data structure associated with a basic node</span>

<span class="sd">    Attributes:</span>
<span class="sd">        id (int): the global node id</span>
<span class="sd">        cid (int): the context node id</span>
<span class="sd">        children (list of int): the list of child nodes in terms of global node ids</span>
<span class="sd">        parents (list of int): the list of parent nodes in terms of global node ids</span>
<span class="sd">        n_parents (int): the number of parent nodes</span>
<span class="sd">        n_children (int): the number of children nodes</span>
<span class="sd">        n_ancestors (int): the number of ancestral (recursive parent) nodes</span>
<span class="sd">        n_descendents (int): the number of descendent (recursive children) nodes</span>
<span class="sd">        root (bool): whether or not the node is a root</span>
<span class="sd">        leaf (bool): whether or not the node is a leaf</span>

<span class="sd">    &quot;&quot;&quot;</span>

<span class="k">class</span> <span class="nc">Node</span><span class="p">(</span><span class="n">BaseNode</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Graph layout node which inherits the BaseNode object.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        name (str): the name annotation of the node, such as &quot;GO:....&quot;</span>
<span class="sd">        weight (int): the number of gene annotations</span>
<span class="sd">        depth (int): the longest distance from the roots</span>
<span class="sd">        height (int): the longest distance from the leaves</span>
<span class="sd">        root_distance (int):</span>
<span class="sd">        depth_order: int</span>
<span class="sd">        height_order: int</span>
<span class="sd">        flex: int</span>
<span class="sd">        flex_order: int</span>
<span class="sd">        queried: bool</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">get_position_info</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Create a dictionary of depth, height and flex x,y coordinates</span>

<span class="sd">        Args:</span>
<span class="sd">            None</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: the dictionary with x,y, coordinates under different views</span>

<span class="sd">        &quot;&quot;&quot;</span>

<span class="k">class</span> <span class="nc">Link</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Data structure for a link</span>

<span class="sd">    Attributes:</span>
<span class="sd">        source (:obj:Node): the parent node</span>
<span class="sd">        target (:obj:Node): the child node</span>

<span class="sd">    &quot;&quot;&quot;</span>

<span class="k">class</span> <span class="nc">DAGraph</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Data structure for a directed cyclic graph</span>

<span class="sd">    Attributes:</span>
<span class="sd">        nodes: a list of Node objects</span>
<span class="sd">        links: a list of Link objects</span>
<span class="sd">        roots: a list of node ids</span>
<span class="sd">        leaves: a list of leaf ids</span>
<span class="sd">        name_index_map: a map from a node name to its index in the `nodes` list</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">filter_redundant_nodes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_map</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">roots</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Apply graph refinement to remove redundant nodes</span>

<span class="sd">        Args:</span>
<span class="sd">            node_map: a map of node ids to Node objects (optional)</span>
<span class="sd">            roots: a list of node ids to be the roots (optional)</span>

<span class="sd">        Returns:</span>
<span class="sd">            a map of node ids to Node objects keeping only non-redundant nodes</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">create_map_to_context_nodes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_context</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Create a dictionary for the context nodes given a set of nodes</span>
<span class="sd">            Node objects that can encode the information of the updated</span>
<span class="sd">            height, depth, parents and children in the particular context</span>
<span class="sd">            defined by the nodes under selection. Relationships between nodes</span>
<span class="sd">            are inferred from the original DAG.</span>

<span class="sd">        Args:</span>
<span class="sd">            note_context: a dict of node ids mapping to anything</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict : a map from node ids to new Node objects in the new context</span>
<span class="sd">            list: a list of root nodes in the new context</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">update_intrinsic_levels</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">metric</span><span class="p">,</span> <span class="n">node_map</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Update the hight or depth attributes based on the optional context</span>

<span class="sd">        Args:</span>
<span class="sd">            metric: either &quot;height&quot; or &quot;depth&quot;</span>
<span class="sd">            node_map: a map from node ids to Node objects (optional)</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a map from node ids to Node objects with updated attributes</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">find_nodes_in_context</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_type</span><span class="p">,</span> <span class="n">node_context</span><span class="p">,</span> <span class="n">node_map</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Find either roots or leaves in a node context</span>

<span class="sd">        Args:</span>
<span class="sd">            node_type: either &quot;root&quot; of &quot;leaf&quot;</span>
<span class="sd">            node_context: a dict or set of node ids</span>
<span class="sd">            node_map: a map from node ids to Node Objects (optional)</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: node ids of the roots or the leaves</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">find_neighbors</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                       <span class="n">node_index</span><span class="p">,</span>
                       <span class="n">relation</span><span class="p">,</span>
                       <span class="n">node_map</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span>
                       <span class="n">restrict_set</span><span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Find the neighbors of a node within a context or restricted set</span>

<span class="sd">        Args:</span>
<span class="sd">            node_index: the node id</span>
<span class="sd">            relation: either &quot;parents&quot; or &quot;children&quot;</span>
<span class="sd">            node_map: a map from node ids to Node Objects (optional)</span>
<span class="sd">            restrict_set: a dict or set of node ids to restrict search</span>

<span class="sd">        Returns:</span>
<span class="sd">            a list of (restricted) neighboring node ids</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">relation_search</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                        <span class="n">node_indices</span><span class="p">,</span>
                        <span class="n">relation</span><span class="p">,</span>
                        <span class="n">measure</span> <span class="o">=</span> <span class="s2">&quot;level&quot;</span><span class="p">,</span>
                        <span class="n">node_map</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span>
                        <span class="n">restrict_set</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Perform restricted hierarchical search with a queue</span>

<span class="sd">        Args:</span>
<span class="sd">            node_indices : list of node ids as the anchor for the search</span>
<span class="sd">            relation: either &quot;parents&quot; or &quot;children&quot;</span>
<span class="sd">            measure: only &quot;level&quot; for now</span>
<span class="sd">            node_map: a map from node ids to Node Objects (optional)</span>
<span class="sd">            restrict_set: a dict or set of node ids to restrict search</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: related nodes ids (including the anchors) mapped to measure</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">create_node_grouping</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                             <span class="n">node_list</span><span class="p">,</span>
                             <span class="n">context_map</span><span class="p">,</span>
                             <span class="n">restrict_set</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Given a list of node ids, group their descendants and</span>
<span class="sd">            ancestors into independent subgraphs for better layout.</span>
<span class="sd">            There are always some nodes that are in more than one groups</span>
<span class="sd">            so we want to partition the nodes optimally or greedily</span>

<span class="sd">        Args:</span>
<span class="sd">            node_list: list of node_ids</span>
<span class="sd">            context_map:  a map from node ids to Node Objects (optional)</span>
<span class="sd">            restrict_set: a dict or set of node ids to restrict search</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of set of node ids indicating grouping</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">compute_focus_flex_level</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                 <span class="n">focus_nodes</span><span class="p">,</span>
                                 <span class="n">context_map</span><span class="p">,</span>
                                 <span class="n">gap_break</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span>
                                 <span class="n">block_merge</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Compute the level of each node using the bubble float algorithm</span>
<span class="sd">            heuristic: larger gap_break yields fewer nodes per level</span>

<span class="sd">        Args:</span>
<span class="sd">            focus_nodes: a set (or list or dict) of node ids</span>
<span class="sd">            context_map: a map from node ids to Node Objects</span>
<span class="sd">            gap_break: the maximum node weight difference between two layers</span>
<span class="sd">            block_merge: (default as False for now)</span>

<span class="sd">        Returns:</span>
<span class="sd">            None (only updates the Node objects in the context map)</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">create_level_node_map</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_levels</span><span class="p">,</span> <span class="n">node_indices</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Create a map from level number to the node ids</span>

<span class="sd">        Args:</span>
<span class="sd">            node_levels: a dictionary of node_index to levels (like a context)</span>
<span class="sd">            node_list: a list of nodes to compute the level map for</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: (level) -&gt; [nidx_1, nidx_2, ...] (ordered list)</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">compute_node_order</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_indices</span><span class="p">,</span> <span class="n">level_type</span><span class="p">,</span> <span class="n">context_map</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Given node ids and level type, compute its level specific position</span>
<span class="sd">            Uses the the hierarchical ordering heuristic</span>

<span class="sd">        Args:</span>
<span class="sd">            node_indices: a list/dict/set of node ids</span>
<span class="sd">            level_type: &quot;height&quot;, &quot;depth&quot; or &quot;flex&quot;</span>
<span class="sd">            context_map: a map from node ids to Node Objects with specified attr</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a map from node ids to their order in the level</span>

<span class="sd">        &quot;&quot;&quot;</span>

<span class="k">class</span> <span class="nc">OrderedContext</span><span class="p">():</span>

        <span class="c1"># index all the nodes in a given context cntx_d.</span>
        <span class="c1"># cntx_id is a dictionary mapping full context node id to the node</span>
        <span class="n">sorted_nodes</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">sorted_index_map</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="c1"># fixed_groups: nodes organized in a particular fixed grouping</span>
        <span class="c1"># retrieve the maximum number of levels</span>
        <span class="n">fixed_level_nodes</span> <span class="o">=</span> <span class="bp">None</span>

    <span class="k">def</span> <span class="nf">boundary_histogram_search</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">level_breaks</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        For each level, determine the starting index in that level</span>
<span class="sd">        If the level breaks are [a, b, c], then the counts should be</span>
<span class="sd">        [|{x: x &gt;= a}|, |{x: a &gt; x &gt;= b}|, |{x: b &gt; x &gt;= c}|]</span>

<span class="sd">        NOTE: duplicate boundaries may occur, in this case, this could</span>
<span class="sd">        be due to the fact that there are nodes of the same size in the</span>
<span class="sd">        focus graph. This edge case should be handled before (and possibly</span>
<span class="sd">        after) this funciton is called</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        leve_breaks: obj: `list`</span>
<span class="sd">            A descending list of node weights dividers, e.g., [18529, 75, 1]</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        set</span>
<span class="sd">            dict of node indices mapped to their measure</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">generate_fixed_level_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">lev_t</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">bouyant_context_layout</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">focus_node_ids</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">format_node_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">output_lite_node_info</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">output_plain_tested_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>


<span class="k">class</span> <span class="nc">GOStat</span><span class="p">():</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">set_test_attr_from_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_dict</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">get_test_attr_as_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">reverse_list_to_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_list</span><span class="p">,</span> <span class="n">do_sort</span> <span class="o">=</span> <span class="bp">True</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">setup_simulation_oneway</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">get_simulation_gene_list</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">get_node_meta_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">evaluate_rejections</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rej_list</span><span class="p">,</span> <span class="n">nonnull_type</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">determine_non_null</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">independent_fisher_tests</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_genes</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">independent_binomial_tests</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_genes</span><span class="p">,</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">independent_global_tests</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_pvalues</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;Simes&quot;</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">generate_node_pvals_from_gene_pvals</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                            <span class="n">genes</span><span class="p">,</span>
                                            <span class="n">pvals</span><span class="p">,</span>
                                            <span class="n">test</span><span class="o">=</span><span class="s2">&quot;simes&quot;</span><span class="p">,</span>
                                            <span class="n">cutoff</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                                            <span class="n">verbose</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
        <span class="c1"># the order of the genes should correspond to the p-values</span>
        <span class="c1"># these genes must be entrez ids to match the gene_go_map</span>
        <span class="c1"># returns a list of p-values that correspond to the node ordering</span>
        <span class="c1"># in the Ordered Context</span>


    <span class="k">def</span> <span class="nf">convert_gene_from_to</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">output_ground_truth_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_graph</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>

<span class="k">class</span> <span class="nc">GODAGraph</span><span class="p">(</span><span class="n">DAGraph</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Data structure for the DAG for GO analysis</span>

<span class="sd">    Attributes:</span>
<span class="sd">        cache_dir: directory to store cache files</span>
<span class="sd">        sim_dir: directory to store simulation outputs</span>
<span class="sd">        name: name of the dag</span>
<span class="sd">        go_fname: directory to store the .pkl file for the object</span>
<span class="sd">        ontology: &quot;biological_process&quot;,&quot;cellular_component&quot; or &quot;molecular_function&quot;</span>
<span class="sd">        species: &quot;human&quot; or &quot;mouse&quot;</span>
<span class="sd">        version: time stamp for version control</span>
<span class="sd">        root: the root of the entire ontolgy</span>
<span class="sd">        go_gene_map: map from GO id to a list of entrez gene ids</span>
<span class="sd">        gene_conversion_map: map from entrez gene id to gene symbol</span>
<span class="sd">        go_annotation: map from GO:id to the full GO name</span>
<span class="sd">        gene_go_map: map from entrez gene id to a list of GO ids</span>
<span class="sd">        context_graph: OrderedContext() object</span>
<span class="sd">        context_params: a data structure of context parameters</span>
<span class="sd">        main_statistician: GOStat() object used for statistical testing</span>
<span class="sd">        gohelper: GOHelper() object used for downloading and parsing the GO</span>

<span class="sd">    &quot;&quot;&quot;</span>


    <span class="k">def</span> <span class="nf">setup_full_dag</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                       <span class="n">ontology</span><span class="p">,</span>
                       <span class="n">species</span><span class="p">,</span>
                       <span class="n">version</span><span class="p">,</span>
                       <span class="n">use_cache</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Integrates annotation information to the GO DAG</span>
<span class="sd">            This steps includes reading the annotation files for a particular</span>
<span class="sd">            root DAG and a particular species, and creating the core structure</span>
<span class="sd">            for context and focus selection. It also reads and writes the .pkl</span>
<span class="sd">            files used to load to populate the Object</span>

<span class="sd">        Args:</span>
<span class="sd">            ontology: &quot;biological_process&quot;,&quot;cellular_component&quot; or &quot;molecular_function&quot;</span>
<span class="sd">            species: &quot;human&quot; or &quot;mouse&quot;</span>
<span class="sd">            version: time stamp for version control</span>
<span class="sd">            use_cache : whether or not to store/use a cached version of the DAG</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            None</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">setup_context_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                            <span class="n">rule</span><span class="p">,</span>
                            <span class="n">target_node_list</span><span class="p">,</span>
                            <span class="n">min_w</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                            <span class="n">max_w</span><span class="o">=</span><span class="mi">30000</span><span class="p">,</span>
                            <span class="n">refine_graph</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                            <span class="n">store_context</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Create a context graph based on anchor nodes and rules</span>

<span class="sd">        Args:</span>
<span class="sd">            rule (str): rule to build the context: &quot;waypoint&quot;, &quot;root&quot; or &quot;leaf&quot;</span>
<span class="sd">            target_node_list (list):  a list of node names (GO:..) anchors</span>
<span class="sd">            min_w (int): the minimum number of genes that each node should have</span>
<span class="sd">            max_w (int):: the maximum number of genes that each node should have</span>
<span class="sd">            refine_graph (bool): whether or not to remove redundent GO terms</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            OrderedContext: the Object representing the context</span>
<span class="sd">            dict: the parameters needed to specify the context</span>

<span class="sd">        &quot;&quot;&quot;</span>


    <span class="k">def</span> <span class="nf">setup_focus_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                          <span class="n">query_go_ids</span><span class="p">,</span>
                          <span class="n">ordered_context</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                          <span class="n">rule</span><span class="o">=</span><span class="s2">&quot;waypoint&quot;</span><span class="p">,</span>
                          <span class="n">max_descendents</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                          <span class="n">force_all_descendents</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                          <span class="n">gap_break</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
                          <span class="n">grouped</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                          <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Create a focus graph based on anchor nodes and rules</span>

<span class="sd">        Args:</span>
<span class="sd">            query_go_ids: a list of GO ids</span>
<span class="sd">            ordered_context: the context graph OrderedContext()</span>
<span class="sd">            rule: rule to build the context: &quot;waypoint&quot;, &quot;root&quot; or &quot;leaf&quot;</span>
<span class="sd">            max_descendents: threshold for a node to be prolific</span>
<span class="sd">            force_all_descendents: whether or not to include all descendants</span>
<span class="sd">            gap_break: break parameter for the buoyant layout</span>
<span class="sd">            grouped: whether or not to segregate the terms into groups</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: the data structure of the graph needed for front-send display</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">restore_testing_configuration</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">job_id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Restores the testing configuration for the main_statistician,</span>
<span class="sd">            It also modifies the OrderedContext() based on the meta parameters</span>

<span class="sd">        Args:</span>
<span class="sd">            job_id: the id of the simulation job</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: all the parameters formatted for front-end rendering *</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">output_node_power_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">job_id</span><span class="p">,</span> <span class="n">test</span><span class="p">,</span> <span class="n">multitest</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Output a matrix of node vs regime with node rejection rate</span>

<span class="sd">        Args:</span>
<span class="sd">            job_id: the id of the simulation job</span>
<span class="sd">            test: the method to generate node p-values</span>
<span class="sd">            multitest: the multiple correction method</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a json-like matrix format</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">output_summary_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">job_id</span><span class="p">,</span> <span class="n">test_method</span><span class="p">,</span> <span class="n">adjust_method</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Output the summary statistics used for plotting</span>

<span class="sd">        Args:</span>
<span class="sd">            job_id: the id of the simulation job</span>
<span class="sd">            test_method: the method to generate node p-values</span>
<span class="sd">            adjust_method: the multiple correction method</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a json-like matrix format for plotting</span>

<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">launch_simulation_pipeline</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">job_id</span><span class="p">,</span> <span class="n">cleanup</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">output_non_null_go_terms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_symb_list</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">output_general_info</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">output_context_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">c_graph</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">plot_full_result</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_dir</span><span class="p">):</span>
</pre></div>

<h3 id="serveralgorithmspy">server/algorithms.py</h3>
<div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">count_inverions</span><span class="p">(</span><span class="n">A</span><span class="p">):</span>

<span class="k">def</span> <span class="nf">inverse_cumsum</span><span class="p">(</span><span class="n">z</span><span class="p">):</span>

<span class="k">def</span> <span class="nf">bubble_float_algo</span><span class="p">(</span><span class="n">nodes</span><span class="p">,</span>
                      <span class="n">node_weights</span><span class="p">,</span>
                      <span class="n">node_depths</span><span class="p">,</span>
                      <span class="n">node_parents</span><span class="p">,</span>
                      <span class="n">node_order</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                      <span class="n">gap_break</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                      <span class="n">block_merge</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
</pre></div>

<h1 id="javascript-documentation">Javascript Documentation</h1>
<h3 id="staticjsall_configsjs">static/js/all_configs.js</h3>
<h3 id="staticjsgo_interactionjs">static/js/go_interaction.js</h3>
<h3 id="staticjscommon_utilsjs">static/js/common_utils.js</h3>
<h3 id="staticjsgraph_utilsjs">static/js/graph_utils.js</h3>
<h3 id="staticjsbar_utilsjs">static/js/bar_utils.js</h3>
<h3 id="staticjsss_manhattanjs">static/js/ss_manhattan.js</h3>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../manual/" title="Web Interface Manual" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Web Interface Manual
              </span>
            </div>
          </a>
        
        
          <a href="../notebooks/aegis_single_cell_example.html" title="Single Cell RNA-seq Application" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Single Cell RNA-seq Application
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.583bbe55.js"></script>
      
      <script>app.initialize({version:"1.0",url:{base:".."}})</script>
      
        <script src="../javascripts/extra.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      
    
    
      
    
  </body>
</html>